# Research on the quality of localization of movie titles

## Importing needed packages

In [1]:
# !pip install spacy spacy-langdetect sentence-transformers
# !python -m spacy download en_core_web_sm

In [2]:
import re
from pprint import PrettyPrinter
from typing import List, Tuple

import numpy as np
import pandas as pd
import spacy
import torch
from sentence_transformers import SentenceTransformer, util
from spacy.language import Language
from spacy_langdetect import LanguageDetector
from torch import Tensor

In [3]:
SEED = 42

The goal of this research is to find out 

## Reading data

From the whole dataset with information about films we'll just read the next columns:
* `id` - movie id
* `russian_title` - Russian title
* `original_title` - original title 

In [4]:
movie_df = pd.read_parquet("../data/movies.parquet", columns=["id", "russian_title", "original_title"])
movie_df.shape

(984, 3)

In [5]:
movie_df = movie_df.replace(r"^\s*$", np.nan, regex=True)

In [6]:
movie_df.sample(10, random_state=SEED)

id                                 russian_title  \
613   647671                                  Охота (2012)   
451    45146                        Любовь и голуби (1984)   
731    84049          Рапунцель: Запутанная история (2010)   
436     1721                             Лжец, лжец (1997)   
275  1072788                    Дьявол всегда здесь (2020)   
582      469                      Однажды в Америке (1983)   
707   915112                     Призрачная красота (2016)   
299    81555  Загадочная история Бенджамина Баттона (2008)   
718  1236583                   Прошлой ночью в Сохо (2021)   
494   577673                      Место под соснами (2012)   

                          original_title  
613                               Jagten  
451                                  NaN  
731                              Tangled  
436                            Liar Liar  
275               The Devil All the Time  
582          Once Upon a Time in America  
707                    Collateral Beauty  
299  The Curious Case of Benjamin Button  
718                   Last Night in Soho  
494           The Place Beyond the Pines

We can see that the `russian_title` contains the release year for every film (or at least these 10 random samples).  
This is an opportunity to clean the data.

## Cleaning the data

### Removing parenthesis from `russian_title`

#### Approach 1: Removing last 6 characters

In the previous section we've found out that the `russian_title` contains the release year of the film.  
Let's check that the last six characters from the `russian_title` is always the same and look like `(year)`.

In [7]:
six_chars = movie_df["russian_title"].apply(lambda s: s[-6:].replace("(", "").replace(")", ""))
six_chars.value_counts(ascending=True).iloc[:15]

1968     1
1976     1
1959     1
 1988    1
1956     1
1975     1
 2022    1
1970     1
1939     1
1974     1
1977     1
1971     1
1989     2
1973     2
1993     2
Name: russian_title, dtype: int64

We can see that amongst the most unfrequent years there are little errors ` 1998` and ` 2022` - year contains additional whitespaces.  
I'm going to check the whole title for this case.

In [8]:
indices = [i for i, val in enumerate(six_chars) if " " in val]
movie_df["russian_title"][indices]

705    Привилегированные (ТВ, 2022)
775       Собачье сердце (ТВ, 1988)
Name: russian_title, dtype: object

Aha!

#### Approach 2: removing the whole parenthesis

Let's switch to another strategy - finding out whether every title contains substring like `(smth)` and if it is true, then remove such substring

In [9]:
def has_numbers_in_square_brackets(s):
    return bool(re.search(r"\(.*\)", s))

In [10]:
assert (
    movie_df["russian_title"].apply(has_numbers_in_square_brackets).sum() == movie_df.shape[0]
), "Not every title has brackets with something inside"

Every title contains some information in brackets - we don't really care what's inside them. Our goal is to clean the titles, so, we'll just delete the brackets with their contents.

In [11]:
movie_df["russian_title"] = movie_df["russian_title"].apply(lambda s: re.sub(r"\([^()]*\)", "", s).strip())
movie_df["russian_title"]

0                           1+1
1      10 причин моей ненависти
2                12 лет рабства
3                    12 обезьян
4        12 разгневанных мужчин
                 ...           
979                  Я — начало
980                    Я, робот
981                Яйцо Фаберже
982                      Ярость
983                      Ярость
Name: russian_title, Length: 984, dtype: object

Just checking

In [12]:
assert (
    movie_df["russian_title"].apply(has_numbers_in_square_brackets).sum() == 0
), "Not all brackets were deleted - check the procedure"

### Checking for duplicates

In [13]:
movie_df.drop_duplicates().shape == movie_df.shape

True

There are no duplicates in the dataset - nothing to do here

### Checking for missing values

In [14]:
movie_df.isna().any()

id                False
russian_title     False
original_title     True
dtype: bool

We can see that the 'original_title' column contains `NaN` values.  
Probably, because not every movie has an `original_title` - maybe movie is Russian-made and doesn't have English-translated title, for example.

In [15]:
movie_df[movie_df.isnull().any(axis=1)]

id                      russian_title original_title
8    4533880          1941. Крылья над Берлином            NaN
18     84674                             9 рота            NaN
41      8385                      Андрей Рублев            NaN
53   1291197         Артек. Большое путешествие            NaN
60   4550354                              Бабки            NaN
..       ...                                ...            ...
944  1209750  Чернобыль: Зона отчуждения. Финал            NaN
969  4312912               Этерна: Часть первая            NaN
975  4903616                       Я иду искать            NaN
977  4493006                    Я хочу! Я буду!            NaN
981  4491514                       Яйцо Фаберже            NaN

[131 rows x 3 columns]

Yes, my assumption was right.  
I think it is reasonable to drop such rows.

In [16]:
movie_df = movie_df.dropna(axis=0, inplace=False) if movie_df.isna().any().any() else movie_df
movie_df.shape

(853, 3)

In [17]:
assert movie_df.isna().any().any() == False

## Semantic similarity

### Getting titles

In [18]:
russian_titles = movie_df["russian_title"].values
original_titles = movie_df["original_title"].values

### Useful functions

In [33]:
def get_embeddings(model, texts: List[str]) -> Tensor:
    texts_lowercase = list(map(str.lower, texts))
    with torch.no_grad():
        text_embeddings = model.encode(texts_lowercase, convert_to_tensor=True)
    return text_embeddings

In [34]:
def compute_embeddings(model, russian_titles: np.array, original_titles: np.array) -> Tuple[Tensor]:
    russian_title_embs = get_embeddings(model, russian_titles)
    original_title_embs = get_embeddings(model, original_titles)
    
    return russian_title_embs, original_title_embs

In [35]:
def compute_similarity(russian_title_embs: Tensor, original_title_embs: Tensor) -> Tensor:
    return util.cos_sim(russian_title_embs, original_title_embs).cpu().detach().numpy()

In [42]:
def get_similarity_dataframe(model, russian_titles, original_titles, sort=False, ascending=True):
    embeddings = compute_embeddings(model, russian_titles, original_titles)
    similarity_scores = compute_similarity(*embeddings)

    rows = []
    for i in range(len(russian_titles)):
        rows.append([russian_titles[i], original_titles[i], similarity_scores[i][i]])

    similarity_df = pd.DataFrame(data=rows, columns=["russian_title", "original_title", "similarity"])
    similarity_df["similarity"] = similarity_df["similarity"].apply(lambda similarity: round(similarity, 3))
    
    del embeddings
    torch.cuda.empty_cache()
    
    similarity_df = similarity_df.sort_values(by="similarity", ascending=ascending) if sort else similarity_df
    
    return similarity_df

#### Debug opportunities

In [30]:
# embeddings = compute_embeddings(distil_use_v2, russian_titles, original_titles)
# similarity_scores = compute_similarity(*embeddings)

### Choosing model

We are going to use multilingual models from [SentenceTransformers](https://www.sbert.net/docs/pretrained_models.html#multi-lingual-models) framework

In [24]:
%%time

distil_use_v1 = SentenceTransformer("distiluse-base-multilingual-cased-v1", cache_folder="../cache_folder")
distil_use_v2 = SentenceTransformer("distiluse-base-multilingual-cased-v2", cache_folder="../cache_folder")
minilm = SentenceTransformer("paraphrase-multilingual-MiniLM-L12-v2", cache_folder="../cache_folder")
mpnet = SentenceTransformer("paraphrase-multilingual-mpnet-base-v2", cache_folder="../cache_folder")
labse = SentenceTransformer("LaBSE", cache_folder="../cache_folder")

CPU times: total: 22.8 s
Wall time: 1min 5s


In [25]:
models = {
    "distiluse-base-multilingual-cased-v1": distil_use_v1,
    "distiluse-base-multilingual-cased-v2": distil_use_v2,
    "paraphrase-multilingual-MiniLM-L12-v2": minilm,
    "paraphrase-multilingual-mpnet-base-v2": mpnet,
    "LaBSE": labse,
}

### Sanity check

Before moving on I'd like to check the sanity of the model - assess how well they predict the similarity between Russian and original titles

In [64]:
similarities = {}
for model_name, model in models.items():
    similarity_df = get_similarity_dataframe(model, russian_titles, original_titles)
    similarities[model_name] = similarity_df["similarity"]
    print(model_name)
    with pd.option_context("display.max_rows", None, "display.max_columns", None):
        display(similarity_df.sample(10, random_state=SEED))
    print("\n")

distiluse-base-multilingual-cased-v1


russian_title  \
66                             Бесконечность   
434                            Мемуары гейши   
198                Девочка, покорившая время   
212                                Дивергент   
652                        Святые из Бундока   
543                                Пассажиры   
280  Звёздные войны: Эпизод 2 — Атака клонов   
296                       Зомби по имени Шон   
365                             Крупная рыба   
679                          Список Шиндлера   

                                   original_title  similarity  
66                                       Infinite       0.896  
434                           Memoirs of a Geisha       0.464  
198                           Toki o kakeru shojo       0.152  
212                                     Divergent       0.735  
652                           The Boondock Saints       0.882  
543                                    Passengers       0.956  
280  Star Wars: Episode II - Attack of the Clones       0.715  
296                             Shaun of the Dead       0.627  
365                                      Big Fish       0.965  
679                              Schindler's List       0.878



distiluse-base-multilingual-cased-v2


russian_title  \
66                             Бесконечность   
434                            Мемуары гейши   
198                Девочка, покорившая время   
212                                Дивергент   
652                        Святые из Бундока   
543                                Пассажиры   
280  Звёздные войны: Эпизод 2 — Атака клонов   
296                       Зомби по имени Шон   
365                             Крупная рыба   
679                          Список Шиндлера   

                                   original_title  similarity  
66                                       Infinite       0.841  
434                           Memoirs of a Geisha       0.765  
198                           Toki o kakeru shojo       0.328  
212                                     Divergent       0.697  
652                           The Boondock Saints       0.858  
543                                    Passengers       0.952  
280  Star Wars: Episode II - Attack of the Clones       0.798  
296                             Shaun of the Dead       0.631  
365                                      Big Fish       0.931  
679                              Schindler's List       0.926



paraphrase-multilingual-MiniLM-L12-v2


russian_title  \
66                             Бесконечность   
434                            Мемуары гейши   
198                Девочка, покорившая время   
212                                Дивергент   
652                        Святые из Бундока   
543                                Пассажиры   
280  Звёздные войны: Эпизод 2 — Атака клонов   
296                       Зомби по имени Шон   
365                             Крупная рыба   
679                          Список Шиндлера   

                                   original_title  similarity  
66                                       Infinite       0.954  
434                           Memoirs of a Geisha       0.787  
198                           Toki o kakeru shojo       0.536  
212                                     Divergent       0.612  
652                           The Boondock Saints       0.727  
543                                    Passengers       0.992  
280  Star Wars: Episode II - Attack of the Clones       0.954  
296                             Shaun of the Dead       0.490  
365                                      Big Fish       0.980  
679                              Schindler's List       0.701



paraphrase-multilingual-mpnet-base-v2


russian_title  \
66                             Бесконечность   
434                            Мемуары гейши   
198                Девочка, покорившая время   
212                                Дивергент   
652                        Святые из Бундока   
543                                Пассажиры   
280  Звёздные войны: Эпизод 2 — Атака клонов   
296                       Зомби по имени Шон   
365                             Крупная рыба   
679                          Список Шиндлера   

                                   original_title  similarity  
66                                       Infinite       0.898  
434                           Memoirs of a Geisha       0.838  
198                           Toki o kakeru shojo       0.467  
212                                     Divergent       0.562  
652                           The Boondock Saints       0.728  
543                                    Passengers       0.963  
280  Star Wars: Episode II - Attack of the Clones       0.950  
296                             Shaun of the Dead       0.624  
365                                      Big Fish       0.925  
679                              Schindler's List       0.724



LaBSE


russian_title  \
66                             Бесконечность   
434                            Мемуары гейши   
198                Девочка, покорившая время   
212                                Дивергент   
652                        Святые из Бундока   
543                                Пассажиры   
280  Звёздные войны: Эпизод 2 — Атака клонов   
296                       Зомби по имени Шон   
365                             Крупная рыба   
679                          Список Шиндлера   

                                   original_title  similarity  
66                                       Infinite       0.874  
434                           Memoirs of a Geisha       0.761  
198                           Toki o kakeru shojo       0.279  
212                                     Divergent       0.796  
652                           The Boondock Saints       0.822  
543                                    Passengers       0.967  
280  Star Wars: Episode II - Attack of the Clones       0.892  
296                             Shaun of the Dead       0.437  
365                                      Big Fish       0.942  
679                              Schindler's List       0.863

### Averaging similarities

They seem to work fine, but their results vary. So what if we calculate average similarity between all the models.  
Can this approach give us better overall performance?

In [65]:
similarity_df.drop("similarity", axis=1, inplace=True, errors='ignore')
for model_name, similarity_col in similarities.items():
    similarity_df[model_name] = similarity_col

similarity_df["avg_sim"] = similarity_df[similarities.keys()].mean(axis=1)

In [66]:
with pd.option_context("display.max_rows", None, "display.max_columns", None):
    display(similarity_df.sort_values(by="avg_sim", ascending=True).iloc[:50])

russian_title  \
80                                       Борат   
505                              Одинокий волк   
507                         Однажды в Ирландии   
570                    По соображениям совести   
810               Человек, который изменил всё   
757                       Унесённые призраками   
765                   Философы: Урок выживания   
139                   Воспоминания об убийстве   
51   Атака титанов. Фильм первый: Жестокий мир   
205                                 День курка   
311                             Иллюзия обмана   
449                       Монстры на каникулах   
771                               Форма голоса   
790      Хочу съесть твою поджелудочную железу   
750                           Убойные каникулы   
113                                    Веном 2   
158                     Гангстер, коп и дьявол   
474               Невероятный мир глазами Энцо   
648                  Самый пьяный округ в мире   
487                            Никто не узнает   
420                        Мальчишник в Вегасе   
0                                          1+1   
484                                      Нефть   
85                   Бриллиантовый полицейский   
54                                   Аферистка   
297                             И грянул шторм   
753    Укрась прощальное утро цветами обещания   
695                   Сядь за руль моей машины   
653                               Сезон дождей   
582                Пол: Секретный материальчик   
251                        Её заветное желание   
687                           Стражи терракоты   
773                                   Форсаж 8   
784                              Ходячий замок   
349                         Красавица и дракон   
547                           Первый встречный   
567                                 Плутовство   
448                                    Монстро   
129                              Во всё тяжкое   
270                  Заклятье: Спуск к дьяволу   
496                               Области тьмы   
199                              Девушка брата   
751  Удивительное путешествие доктора Дулиттла   
568                               По наклонной   
106                                     Ведьма   
588                     Полынь: Дорога мёртвых   
768                           Флирт с дьяволом   
846                               Я иду искать   
638                             Реальные упыри   
399                        Любовное настроение   

                                        original_title  \
80   Borat: Cultural Learnings of America for Make ...   
505                                              Clean   
507                                          The Guard   
570                                      Hacksaw Ridge   
810                                          Moneyball   
757                      Sen to Chihiro no kamikakushi   
765                                     After the Dark   
139                                     Salinui chueok   
51                                  Shingeki no kyojin   
205                                         Boss Level   
311                                     Now You See Me   
449                                 Hotel Transylvania   
771                                     Koe no katachi   
790                           Kimi no suizo wo tabetai   
750                            Tucker and Dale vs Evil   
113                        Venom: Let There Be Carnage   
158                                           Akinjeon   
474                      The Art of Racing in the Rain   
648                                            Lawless   
487                                   Dare mo shiranai   
420                                       The Hangover   
0                                         Intouchables   
484                                There Will Be Blood   
85                                         Blue Streak   
54                              

In [60]:
with pd.option_context("display.max_rows", None, "display.max_columns", None):
    display(similarity_df.sort_values(by="avg_sim", ascending=False).iloc[:50])

russian_title                original_title  \
6                          1408                          1408   
7                          1917                          1917   
23                            X                             X   
10                         2012                          2012   
13                     365 дней                       365 dni   
578                      Пожары                     Incendies   
84                       Братья                      Brothers   
515               Опасный метод            A Dangerous Method   
675             Социальная сеть            The Social Network   
607       Преступления будущего          Crimes of the Future   
287               Зеленая книга                    Green Book   
21   Kingsman: Секретная служба  Kingsman: The Secret Service   
818                 Черный ящик                   Boîte noire   
466             Назад в будущее            Back to the Future   
384                     Лето 84                  Summer of 84   
422                       мама!                       mother!   
700                       Такси                          Taxi   
142                        Враг                         Enemy   
543                   Пассажиры                    Passengers   
805          Чего хотят женщины               What Women Want   
708                        Тело                     El cuerpo   
421                        Мама                          Mama   
97                В сердце моря       In the Heart of the Sea   
362               Кровь и кость                Blood and Bone   
788                 Хороший год                   A Good Year   
802              Чёрный телефон               The Black Phone   
203               Девять жизней                    Nine Lives   
579                       Поиск                     Searching   
336                       Код 8                        Code 8   
338                     Контакт                       Contact   
499     Общество мертвых поэтов            Dead Poets Society   
710                 Темные воды                   Dark Waters   
415           Маленькие женщины                  Little Women   
175               Глубокие воды                    Deep Water   
655             Семь психопатов             Seven Psychopaths   
486                       Никто                        Nobody   
227                   Дом Gucci                House of Gucci   
603                  Почему он?                      Why Him?   
691                       Судья                     The Judge   
148             Все старые ножи            All the Old Knives   
469            Не время умирать                No Time to Die   
502                   Один день                       One Day   
723                 Тихое место                 A Quiet Place   
698           Тайна в их глазах          Secret in Their Eyes   
365                Крупная рыба                      Big Fish   
478                 Неизвестный                       Unknown   
4        12 разгневанных мужчин                  12 Angry Men   
375                     Легенда                        Legend   
630               Пятый элемент             The Fifth Element   
224  Дождливый день в Нью-Йорке       A Rainy Day in New York   

     distiluse-base-multilingual-cased-v1  \
6                                   1.000   
7                                   1.000   
23                                  1.000   
10                                  1.000   
13                                  0.993   
578                                 0.956   
84                                  0.981   
515                                 0.969   
675                                 0.962   
607                                 0.962   
287                                 0.969   
21                                  0.980   
818                                 0.946   
466                                 0.958   
384                        

### Creating resulting DataFrame

In [103]:
similarity_df = get_similarity_dataframe(model, russian_titles, original_titles)
similarity_df.head()

russian_title              original_title  similarity
0                       1+1                Intouchables       0.190
1  10 причин моей ненависти  10 Things I Hate About You       0.791
2            12 лет рабства            12 Years a Slave       0.939
3                12 обезьян              Twelve Monkeys       0.932
4    12 разгневанных мужчин                12 Angry Men       0.962

## Results

### Dissimilar titles

Let's look at titles for which Russian translations doesn't convey meaning of the original name. 

In [77]:
with pd.option_context("display.max_rows", None, "display.max_columns", None):
    display(similarity_df.sort_values(by="similarity", ascending=True).iloc[:50])

russian_title  \
80                                       Борат   
765                   Философы: Урок выживания   
507                         Однажды в Ирландии   
113                                    Веном 2   
311                             Иллюзия обмана   
474               Невероятный мир глазами Энцо   
810               Человек, который изменил всё   
169                  Гарри Хафт: Последний бой   
751  Удивительное путешествие доктора Дулиттла   
505                              Одинокий волк   
55                                     Бёрдмэн   
790      Хочу съесть твою поджелудочную железу   
648                  Самый пьяный округ в мире   
270                  Заклятье: Спуск к дьяволу   
588                     Полынь: Дорога мёртвых   
784                              Ходячий замок   
582                Пол: Секретный материальчик   
638                             Реальные упыри   
635              Рапунцель: Запутанная история   
106                                     Ведьма   
0                                          1+1   
82                            Босс-молокосос 2   
42             Анчартед: На картах не значится   
349                         Красавица и дракон   
484                                      Нефть   
547                           Первый встречный   
297                             И грянул шторм   
449                       Монстры на каникулах   
773                                   Форсаж 8   
215                                Дитя погоды   
51   Атака титанов. Фильм первый: Жестокий мир   
220               Добро пожаловать в Zомбилэнд   
779                   Хинтерленд: город грехов   
846                               Я иду искать   
496                               Области тьмы   
205                                 День курка   
750                           Убойные каникулы   
85                   Бриллиантовый полицейский   
17              CODA: Ребенок глухих родителей   
625          Психиатрическая больница Конджиам   
420                        Мальчишник в Вегасе   
487                            Никто не узнает   
129                              Во всё тяжкое   
570                    По соображениям совести   
753    Укрась прощальное утро цветами обещания   
151                           Вторая жизнь Уве   
695                   Сядь за руль моей машины   
199                              Девушка брата   
95                В лес, где мерцают светлячки   
520                              Особое мнение   

                                        original_title  similarity  
80   Borat: Cultural Learnings of America for Make ...       0.033  
765                                     After the Dark       0.075  
507                                          The Guard       0.104  
113                        Venom: Let There Be Carnage       0.112  
311                                     Now You See Me       0.127  
474                      The Art of Racing in the Rain       0.130  
810                                          Moneyball       0.130  
169                                       The Survivor       0.141  
751                                           Dolittle       0.142  
505                                              Clean       0.144  
55     Birdman or (The Unexpected Virtue of Ignorance)       0.146  
790                           Kimi no suizo wo tabetai       0.158  
648                                            Lawless       0.160  
270                                         The Cellar       0.163  
588                                           Wyrmwood       0.164  
784                               Hauru no ugoku shiro       0.166  
582                                               Paul       0.169  
638                          What We Do in the Shadows       0.169  
635                                            Tangled       0.181  
106                 The VVitch: A New-England Folktale       0.188  
0                                

There are a few cases for dissimilarity:  

**Russian title is a cropped version of original title**  
Another problem in this case can be the fact that embeddings don't work very well with proper names like Borat::Борат, Dolittle::Дулиттл, and so on.  
Examples:
* Борат::Borat: Cultural Learnings of America for Make Benefit Glorious Nation of Kazakhstan (sim score: 0.033)
* Веном 2::Venom: Let There Be Carnage (sim score: 0.112)
* Бёрдмэн::Birdman or (The Unexpected Virtue of Ignorance) (sim score: 0.146)

**Russian title is an extended version of original title**  
(Remark about proper names applies to this case too)  
Examples:
* Удивительное путешествие доктора Дулиттла::Dolittle (sim score: 0.142)
* Пол: Секретный материальчик::Paul (sim score: 0.169)
* Рапунцель: Запутанная история::Tangled (sim score: 0.181)

**Russian title was localized (made up) by translators/localizers**  
Examples:
* Невероятный мир глазами Энцо::The Art of Racing in the Rain (sim score: 0.130)
* Человек, который изменил всё::Moneyball (sim score: 0.130)
* Области тьмы::Limitless (sim score: 0.230)

**Errors due to the language of the original title**  
Examples:
* Хочу съесть твою поджелудочную железу::Kimi no suizo wo tabetai (sim score: 0.158)
* Красавица и дракон::Ryu to Sobakasu no Hime (sim score: 0.199)
* Дитя погоды::Tenki no k (sim score: 0.218)

### Similar titles

In [78]:
with pd.option_context("display.max_rows", None, "display.max_columns", None):
    display(similarity_df.sort_values(by="similarity", ascending=False).iloc[:30])

russian_title                original_title  similarity
23                            X                             X       1.000
6                          1408                          1408       1.000
7                          1917                          1917       1.000
10                         2012                          2012       1.000
13                     365 дней                       365 dni       0.997
708                        Тело                     El cuerpo       0.990
578                      Пожары                     Incendies       0.989
421                        Мама                          Mama       0.989
466             Назад в будущее            Back to the Future       0.982
21   Kingsman: Секретная служба  Kingsman: The Secret Service       0.982
579                       Поиск                     Searching       0.977
502                   Один день                       One Day       0.977
287               Зеленая книга                    Green Book       0.975
384                     Лето 84                  Summer of 84       0.974
486                       Никто                        Nobody       0.973
338                     Контакт                       Contact       0.973
375                     Легенда                        Legend       0.972
539         Параллельные матери              Madres paralelas       0.971
84                       Братья                      Brothers       0.971
422                       мама!                       mother!       0.970
499     Общество мертвых поэтов            Dead Poets Society       0.970
802              Чёрный телефон               The Black Phone       0.968
607       Преступления будущего          Crimes of the Future       0.968
675             Социальная сеть            The Social Network       0.968
336                       Код 8                        Code 8       0.968
353      Красота по-американски               American Beauty       0.967
655             Семь психопатов             Seven Psychopaths       0.966
415           Маленькие женщины                  Little Women       0.965
188                  Гравитация                       Gravity       0.964
97                В сердце моря       In the Heart of the Sea       0.962

Similar titles are easier - they are almost literal translation of original titles. 

## Things to do

* Visualization:
    * [Visualizing Embeddings With t-SNE](https://www.kaggle.com/code/colinmorris/visualizing-embeddings-with-t-sne/notebook)